lhs.js

```diff
@@ -48,8 +48,10 @@ class Lhs {
     }
 
     when(pred) {
-        this._predicate = pred;
-        return this;
+        const selector = this._type !== undefined? this._type : this.selector;
```

Comment:
So Lhs doesn't have to know about its subclasses (which is a bad pattern in general because it's a circular dependency: you can never make a new subclass without modifying the superclass), why not introduce a clone() method that TypeLhs and DomLhs can implement and Lhs can specify? Otherwise, this looks good as far as immutability goes.